common4j/src/main/com/microsoft/identity/common/java/interfaces/IPerSeparatorMultiTypeNameValueStore.java

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.interfaces;
+
+import com.microsoft.identity.common.java.util.ported.Predicate;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+import lombok.NonNull;
+
+/**
+ * This is a storage that allows to store name-value pairs associated to a particular "Separator".
+ * This separator can be anything of type T as defined by the implementations of this interface.
+ * <p>
+ * Functionally, this interface should be considered similar to Account Manager in Android that
+ * allows to store name-value pairs separated per Account. This interface does the same thing,
+ * however, instead of separating per account, it allows to separate per anything of any type.
+ * Furthermore, internal storage mechanism here can also be backed by anything i.e. this interface
+ * can be implemented by SharedPreferences on Android or any other mechanism on any platform.
+ *
+ * @param <T> the type of separator to use
+ */
+public interface IPerSeparatorMultiTypeNameValueStore<T> {
+    /**
+     * Gets a {@link String} value from the storage.
+     *
+     * @param separator the key to use for storage separation
+     * @param name      A name associated to the value
+     */
+    @Nullable
+    String getString(@NonNull T separator, @NonNull String name);
```

Comment:
what would be returned if the provided separator is not found?
(we probably do not want this to vary for each implementation).

Reply:
asking because... if the developer should know t

Reply:
and if that matters, the alternative is to

1. Create IKeyValueStorage<U,T> as a base class of INameValueStorage<String, T>
2. Whenever we need this, we create a nested INameValueStorage<Separator, INameValueStorage<String, T>>

Reply:
Should be getting back `null`.

This whole class is actually almost the same as [IAccountDataStorage](https://github.com/AzureAD/ad-accounts-for-android/blob/dev/broker4j/src/main/com/microsoft/identity/broker4j/broker/platform/components/IAccountDataStorage.java#L301-L321)

that one seem to not document what would happen if `Account` isn't found but from the look of the code it seems to return back null. The class added in this PR will do the same thing.